### PR TITLE
[PHP] Pedantic, Explicit Type Enforcement

### PR DESCRIPTION
**Pull request by paragonie-scott**

### CODE REVIEW

**Comment by paragonie-scott** [comment]
Similar to #61

File: PasswordStorage.php [hunk 1]
```diff
@@ -19,10 +19,25 @@ class PasswordStorage {
...
+            throw new InvalidArgumentException(
+                "create_hash(): Expected a string"
+            );
+        }
+        if (\function_exists('random_bytes')) {
+            $salt_raw = \random_bytes(self::PBKDF2_SALT_BYTES);
+        } else {
+            $salt_raw = \mcrypt_create_iv(self::PBKDF2_SALT_BYTES, MCRYPT_DEV_URANDOM);
+        }
         if ($salt_raw === false) {
```
Comment (defuse) [inline PasswordStorage.php]: `random_bytes` doesn't return `false` on error, it throws exceptions. We kinda promise to transform all exceptions into one of our two exception types (which is kinda stupid now that I think of it). So we have to either give up on that promise or put a try-catch around `random_bytes` that'll convert it into a `CannotPerformOperationException`.

Comment (paragonie-scott) [inline PasswordStorage.php]: Easily done. :)

File: PasswordStorage.php [hunk 2]
```diff
@@ -99,13 +126,28 @@ public static function verify_password($password, $hash)
...
     public static function slow_equals($a, $b)
     {
+        if (!\is_string($a) || !\is_string($b)) {
+            throw new InvalidArgumentException(
+                "slow_equals(): expected two strings"
+            );
+        }
+        if (\function_exists('hash_equals')) {
+            return \hash_equals($a, $b);
+        }
```
Comment (defuse) [inline PasswordStorage.php]: +1

File: PasswordStorage.php [hunk 3]
```diff
@@ -127,8 +169,28 @@ public static function slow_equals($a, $b)
...
+            );
+        }
+        if (!\is_string($salt)) {
+            throw new InvalidArgumentException(
+                "pbkdf2(): salt must be a string"
+            );
+        }
+        // Coerce strings to integers with no information loss or overflow
+        $count += 0;
+        $key_length += 0;
```
Comment (defuse) [inline PasswordStorage.php]: Cool!

File: PasswordStorage.php (hunk 1 above)
Comment (paragonie-scott) [inline PasswordStorage.php]: (And done, of course.)

**Comment by defuse** [comment]
Awesome work!

**Pull request merged**